### PR TITLE
Fix #394 added pkcompat.zip_strict

### LINKED ISSUE #394: pkcompat.zip_strict

### PR DESCRIPTION
**Pull request by robnagler**

### CODE REVIEW

File: pykern/pkcompat.py [hunk 1]
```diff
@@ -99,3 +114,35 @@ def _assert_type(value, typ):
...
+
+
+def _zip_strict(iterables):
+    """Code from https://stackoverflow.com/a/69485272"""
+
+    def _first_tail():
+        nonlocal first_stopped
+        first_stopped = True
+        return
+        yield
```
Comment (e-carlin) [inline pykern/pkcompat.py:133]: How is this reached if we return before?

File: pykern/pkcompat.py [hunk 2]
```diff
@@ -99,3 +114,35 @@ def _assert_type(value, typ):
...
+
+    def _zip_tail():
+        if not first_stopped:
+            raise ValueError("first iterable is longer")
+        for _ in itertools.chain.from_iterable(rest):
+            raise ValueError("first iterable is shorter")
+            yield
```
Comment (e-carlin) [inline pykern/pkcompat.py:140]: How is this reached if we raise before?

File: pykern/pkcompat.py [hunk 3]
```diff
@@ -99,3 +114,35 @@ def _assert_type(value, typ):
                 type(value),
             ),
         )
+
+
+def _version_ok(major, minor):
```
Comment (e-carlin) [inline pykern/pkcompat.py:119]: Is there a reason to prefer this over the code below? I've seen more of the code below (not that that means much...)
```python
def _version_ok(major, minor)
    return sys.version_info >= (major, minor)
```

Comment (robnagler) [reply pykern/pkcompat.py:119]: Thanks. I didn't know what I was doing.

File: pykern/pkcompat.py (hunk 2 above)
Comment (robnagler) [reply pykern/pkcompat.py:140]: The yield just forces it to be a coroutine. It's some pretty tricky code.

File: pykern/pkcompat.py (hunk 1 above)
Comment (robnagler) [reply pykern/pkcompat.py:133]: Yield forces it to be a coroutine (generator).

File: pykern/pkcompat.py (hunk 3 above)
Comment (robnagler) [reply pykern/pkcompat.py:119]: 806acaa

**Comment by robnagler** [comment]
I spent a little time trying to figure out how this magic little function works:

```py
def _zip_strict(iterables):
    """Code from https://stackoverflow.com/a/69485272"""

    def _first_tail():
        nonlocal first_stopped
        first_stopped = True
        return
        yield

    def _zip_tail():
        if not first_stopped:
            raise ValueError("first iterable is longer")
        for _ in itertools.chain.from_iterable(rest):
            raise ValueError("first iterable is shorter")
            yield

    if len(iterables) < 2:
        return zip(*iterables)
    first_stopped = False
    iterables = iter(iterables)
    first = itertools.chain(next(iterables), _first_tail())
    rest = list(map(iter, iterables))
    return itertools.chain(zip(first, *rest), _zip_tail())
```

Some definitions:
- `next(any_iterator)` must return `StopIteration` after it was already raised
- `iter` on an iterator returns self
- `zip` doesn't require an iterator 
- `zip` stops iterating the rest of its arguments once an argument is exhausted. This is similar to `in`, btw.
- `chain` concatenates a sequence of iterables
- `from_iterable` concatenates  a single argument of iterables 

Algorithm:
- `first` has a "watcher" (`_first_tail`) that is an empty generator (forced by the `yield`) 
- `rest` has no watchers
- `zip` has a watcher (`_zip_tail`) that checks:
  -  if `_first_tail` did not execute, then `rest` was shorter than `iterable[0]`
  - Otherwise, check if there's nothing left in `rest`, that is, `next` on each element of `rest` must return `StopIteration`

Notes:
- It's important that `rest` be iterators, not iterables. That was confusing to me. Only iterators report `StopIteration`, not sequences. 
- `zip` does not care if something is a sequence or an iterator. It will call `iter` on all its arguments. 
- I didn't understand why the `map` had to call `iter`: it's to ensure that `from_iterable`, which also calls `iter`, doesn't start from the beginning. 
- The algorithm above is highly efficient, because `iter` returns self on an iterator. It's a no-op, and itertools is written in C.
- All the work is done by zip until the last iteration when `_first_tail` and `_zip_tail` execute once. No matter how big the iterables are, the code will run almost linearly (as shown [in the graph](Code from https://stackoverflow.com/a/69485272)).

HTH.

File: pykern/pkcompat.py (hunk 1 above)
Comment (e-carlin) [reply pykern/pkcompat.py:130]: Oh very interesting. Thanks.

File: pykern/pkcompat.py [hunk 4]
```diff
@@ -8,11 +8,12 @@
 :license: http://www.apache.org/licenses/LICENSE-2.0.html
```
Comment (rorour) [inline pykern/pkcompat.py:8]: Line 2: `Python version compatibility routines` seems like a more accurate description now

**Review by rorour — APPROVED** [review]
Interesting

File: pykern/pkcompat.py (hunk 4 above)
Comment (robnagler) [reply pykern/pkcompat.py:12]: a2f9f4e2ba9faf60695c181aba2baa216a6cc1f3

**Pull request merged**